Userland/Libraries/LibVideo/VP9/Decoder.cpp

```diff
@@ -152,24 +215,1115 @@ u8 Decoder::adapt_prob(u8 prob, u8 counts[2])
+        }
+    }
+
+    // 2. Dequant[ 0 ][ 0 ] is set equal to ( Tokens[ 0 ] * get_dc_quant( plane ) ) / dqDenom
+    dequantized[0] = (m_parser->m_tokens[0] * get_dc_quant(plane)) / dq_denominator;
+
+    // It is a requirement of bitstream conformance that the values written into the Dequant array in steps 1 and 2
+    // are representable by a signed integer with 8 + BitDepth bits.
+    for (auto i = 0u; i < block_size * block_size; i++)
+        VERIFY(check_intermediate_bounds(dequantized[i]));
+
+    // 3. Invoke the 2D inverse transform block process defined in section 8.7.2 with the variable n as input.
+    //    The inverse transform outputs are stored back to the Dequant buffer.
+    TRY(inverse_transform_2d(dequantized, log2_of_block_size));
+
+    // 4. CurrFrame[ plane ][ y + i ][ x + j ] is set equal to Clip1( CurrFrame[ plane ][ y + i ][ x + j ] + Dequant[ i ][ j ] )
+    //    for i = 0..(n0-1) and j = 0..(n0-1).
+    auto& current_buffer = get_output_buffer(plane);
+    auto subsampling_x = (plane > 0 ? m_parser->m_subsampling_x : 0);
+    auto subsampling_y = (plane > 0 ? m_parser->m_subsampling_y : 0);
+    auto frame_width = (m_parser->m_mi_cols * 8) >> subsampling_x;
+    auto frame_height = (m_parser->m_mi_rows * 8) >> subsampling_y;
+    auto width_in_frame_buffer = min(block_size, frame_width - transform_block_x);
+    auto height_in_frame_buffer = min(block_size, frame_height - transform_block_y);
+
+    for (auto i = 0u; i < height_in_frame_buffer; i++) {
+        for (auto j = 0u; j < width_in_frame_buffer; j++) {
+            auto index = index_from_row_and_column(transform_block_y + i, transform_block_x + j, frame_width);
+            auto dequantized_value = dequantized[index_from_row_and_column(i, j, block_size)];
+            current_buffer[index] = clip_1(m_parser->m_bit_depth, current_buffer[index] + dequantized_value);
+        }
+    }
+
+    return {};
+}
+
+inline DecoderErrorOr<void> Decoder::inverse_walsh_hadamard_transform(Vector<Intermediate>& data, u8 log2_of_block_size, u8 shift)
+{
+    (void)data;
+    (void)shift;
+    // The input to this process is a variable shift that specifies the amount of pre-scaling.
+    // This process does an in-place transform of the array T (of length 4) by the following ordered steps:
+    if (1 << log2_of_block_size != 4)
+        return DECODER_ERROR_CORRUPTED("Block size was not 4");
+
     return DECODER_ERROR_NOT_IMPLEMENTED();
 }
 
+inline i32 Decoder::cos64(u8 angle)
+{
+    const i32 cos64_lookup[33] = { 16384, 16364, 16305, 16207, 16069, 15893, 15679, 15426, 15137, 14811, 14449, 14053, 13623, 13160, 12665, 12140, 11585, 11003, 10394, 9760, 9102, 8423, 7723, 7005, 6270, 5520, 4756, 3981, 3196, 2404, 1606, 804, 0 };
+
+    // 1. Set a variable angle2 equal to angle & 127.
+    angle &= 127;
+    VERIFY(angle > 0);
+    // 2. If angle2 is greater than or equal to 0 and less than or equal to 32, return cos64_lookup[ angle2 ].
```

Comment:
I think that VERIFY may have been from when the angle value was signed, although it should have been `angle >= 0` in the first place, I believe. Been a while since I wrote this code so I forget if there are any cases where the angle isn't a fixed value.

I'll remove the VERIFY.